### PR TITLE
[AWS Bedrock] Add Field to hold LLM text model response value in human readable format for selected LLM models

### PR DESCRIPTION
**Pull request by agithomas**
## Enhancement

Create a new field with the field to store the concatenated stream data in `gen_ai.completion` field, when LLM text models are used. A new dashboard panel is introduced to show the LLM prompt and  LLM prompt response along with the used input and output tokens. Added dashboard panel filters for AWS Account Name, AWS Region

## Proposed commit message

Create a new field with the field to store the concatenated stream data in `gen_ai.completion` field, when LLM text models are used.  A new dashboard panel is introduced to show the LLM prompt and  LLM prompt response along with the used input and output tokens. Added dashboard panel filters for AWS Account Name, AWS Region

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

Tested with following AWS Foundation models

Titan
Anthropic Claude
Anthropic Haiku
Llama3
Mistral
Titan Text G1 - Express

## Notes

AWS Bedrock can host custom models. In the event that the `gen_ai.completion` field value is different from one of the above mentioned models, the value of the field will be empty string.

## How to test this PR locally

- elastic-package build
- elastic-package stack up -v -d --services package-registry
- Make use of AWS Bedrock playground, one of the above mentioned models, run the prompt.

## Related issues

- Closes https://github.com/elastic/integrations/issues/10875
- Closes https://github.com/elastic/integrations/issues/10873

## Screenshots

image

image

### CODE REVIEW

**Comment by muthu-mps** [comment]
The logs dashboard uses the `gen_ai.*` fields for prompt text and token_usage and for the response completion text the `aws.bedrock.invocation.*` field is used. Is it possible to ingest the response text in the gen_ai.* field?

**Comment by agithomas** [comment]
> The logs dashboard uses the `gen_ai.*` fields for prompt text and token_usage and for the response completion text the `aws.bedrock.invocation.*` field is used. Is it possible to ingest the response text in the gen_ai.* field?

After referring to this [comment](https://github.com/elastic/integrations/pull/9667#pullrequestreview-2037126807), field names beginning with `gen_ai` are proposed OTEL fields. 

Since I could not find an equivalent name in the proposed list, i kep the field name prefix as `aws_bedrock.invocation.output` instead of `gen_ai.`

```

File: packages/aws_bedrock/changelog.yml [hunk 1]
```diff
@@ -1,3 +1,8 @@
+- version: "0.6.0"
+  changes:
+    - description: Add new field having LLM text model response. Add visualization for LLM prompt and response.
```
Comment (andrewkroh) [inline packages/aws_bedrock/changelog.yml:3]: I suggest mentioning the name of the new field so users know.

```suggestion
    - description: Add new field `aws_bedrock.invocation.output.completion_text` having LLM text model response. Add visualization for LLM prompt and response.
```

File: packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -442,6 +442,78 @@ processors:
...
+            }
+          }
+
+          // Ensure ctx.aws_bedrock.invocation.output is initialized
+          if (ctx.aws_bedrock?.invocation?.output == null) {
+              ctx.aws_bedrock.invocation.output = new HashMap();
+          }
+          // Trim completionText if it exceeds 32766 characters
+          if (completionText.length() > 32766) {
+            completionText.setLength(32766);
```
Comment (muthu-mps) [inline packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml:505]: - We try to trim the characters above `32766`. Is there any difference in using the `ignore_above` value set in the fields.yml?
- I am not sure how we define value for `ignore_above` as we would also get non-ascii characters in the response.

Comment (muthu-mps) [reply packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml:505]: `ignore_above` skips the field itself when exceeding the limit. By this way we can have the trimmed version of the response.

File: packages/aws_bedrock/img/aws_bedrock_chat_image.png
Comment (harnish-elastic) [inline packages/aws_bedrock/img/aws_bedrock_chat_image.png:1]: Added image is not containing Dataset filter. Can we update the screenshot?

File: packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -442,6 +442,78 @@ processors:
...
+  - script: 
+      description: Convert chat completion response in map or list format to user friendly format
+      lang: painless
+      source: |
+        try {
+          def completionText = new StringBuilder();
+          if (ctx.aws_bedrock?.invocation?.output?.output_body_json != null) {
+            if (ctx.aws_bedrock.invocation.output.output_body_json instanceof List) {
+              for (def block : ctx.aws_bedrock.invocation.output.output_body_json) {
+                if (block instanceof Map) {
```
Comment (efd6) [inline packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml:455]: I'd like it if this were inverted and we `continue`d in the case that the condition fires. This reduces the indent level. Similarly in the major else block below.

File: packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    "managed": false,
+    "references": [
+        {
+            "id": "logs-*",
+            "name": "kibanaSavedObjectMeta.searchSourceJSON.index",
+            "type": "index-pattern"
+        }
+    ],
+    "type": "search",
+    "typeMigrationVersion": "10.2.0"
```
Comment (muthu-mps) [inline packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json:45]: The "typeMigrationVersion": "8.9.0" in the dashboard file and it is `10.2.0` is search. I hope this doesn't break things with minimum supported version.

Comment (agithomas) [reply packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json:45]: The dashboard is exported with `8.13.4` version.

Comment (agithomas) [reply packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json:45]: No issues was observed during installing the package on 8.13.4 stack

File: packages/aws_bedrock/img/aws_bedrock_chat_image.png
Comment (agithomas) [reply packages/aws_bedrock/img/aws_bedrock_chat_image.png:1]: Updated the screenshot.

File: packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml (hunk 3 above)
Comment (agithomas) [reply packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml]: Thanks. I have made the change. Not sure, if i understood incorrectly,  in the else clause, as there is no iteration, `continue` cannot be used. 

Please correct me, if i misunderstood.

File: packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json
Comment (harnish-elastic) [inline packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json:1]: AWS Bedrock is using kibana `8.13.0` version as well format version is `3.0.2`, so you can unlink saved search from library and export it (make sure to delete this file). Can you do that here?

File: packages/aws_bedrock/kibana/search/aws_bedrock-102e7881-0c3f-4996-91e6-453f682df2f2.json
Comment (harnish-elastic) [inline packages/aws_bedrock/kibana/search/aws_bedrock-102e7881-0c3f-4996-91e6-453f682df2f2.json:1]: Same here, we can remove/delete this file after exporting.

File: packages/aws_bedrock/data_stream/invocation/_dev/test/pipeline/test-aws-bedrock.log-expected.json
Comment (harnish-elastic) [inline packages/aws_bedrock/data_stream/invocation/_dev/test/pipeline/test-aws-bedrock.log-expected.json:1]: Strange: Without doing a log addition/updation in the `test-aws-bedrock.log` file pipeline tests generated?

**Review by harnish-elastic — CHANGES_REQUESTED** [review]

File: packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml (hunk 3 above)
Comment (efd6) [reply packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml]: Sorry, you are correct.

File: packages/aws_bedrock/data_stream/invocation/_dev/test/pipeline/test-aws-bedrock.log-expected.json
Comment (agithomas) [reply packages/aws_bedrock/data_stream/invocation/_dev/test/pipeline/test-aws-bedrock.log-expected.json:1]: Please refer here.

https://github.com/elastic/integrations/tree/main/packages/aws_bedrock/data_stream/invocation/_dev/test/pipeline

File: packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json
Comment (agithomas) [reply packages/aws_bedrock/kibana/search/aws_bedrock-a8901990-aaf9-469e-aec7-5e6c93d820dc.json:1]: Resolved

File: packages/aws_bedrock/kibana/search/aws_bedrock-102e7881-0c3f-4996-91e6-453f682df2f2.json
Comment (agithomas) [reply packages/aws_bedrock/kibana/search/aws_bedrock-102e7881-0c3f-4996-91e6-453f682df2f2.json:1]: Resolved

**Review by harnish-elastic — APPROVED** [review]

**Review by muthu-mps — APPROVED** [review]
LGTM!

File: packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -442,6 +442,78 @@ processors:
           ctx.aws_bedrock.invocation.input.remove("input_body_json");
           ctx.gen_ai.remove("prompt");
         }
+
+  - script: 
+      description: Convert chat completion response in map or list format to user friendly format
+      lang: painless
+      source: |
```
Comment (efd6) [inline packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml:449]: The script can be simplified considerably

```diff
diff --git a/packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml b/packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml
index b650cda473..3d89c4e6b8 100644
--- a/packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml
+++ b/packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml
@@ -457,41 +457,31 @@ processors:
                 if (block.containsKey('delta') && block.delta instanceof Map && block.delta.containsKey('text')) {
                   // Titan model response
                   completionText.append(block.delta.text);
-                } else if (block.containsKey('outputs') && block.outputs instanceof List) {
+                } else if (block.outputs instanceof List) {
                   // Mistral model response
                   for (def cont : block.outputs) {
-                    if (cont instanceof Map && cont.containsKey('text')) {
+                    if (cont.text instanceof String) {
                         completionText.append(cont.text);
                     }
                   }
-                } else if (block.containsKey('generation')) {
+                } else if (block.generation != null) {
                   // Llama3 model response
                   completionText.append(block.generation);
-                } else if (block.containsKey('outputText')) {
-                    // Titan Text G1 - Express response
-                    completionText.append(block.outputText);
+                } else if (block.outputText != null) {
+                  // Titan Text G1 - Express response
+                  completionText.append(block.outputText);
                 }
               }
-            }else if (ctx.aws_bedrock.invocation.output.output_body_json instanceof Map) {
+            } else if (ctx.aws_bedrock.invocation.output.output_body_json instanceof Map) {
               def block = ctx.aws_bedrock.invocation.output.output_body_json;
-              if (block instanceof Map) {
-                Map blockMap = (Map) block;
-                if (blockMap.containsKey('output') && blockMap.output instanceof Map) {
-                  Map outputMap = (Map) blockMap.output;
-                  if (outputMap.containsKey('message') && outputMap.message instanceof Map) {
-                    Map messageMap = (Map) outputMap.message;
-                    if (messageMap.containsKey('content') && messageMap.content instanceof List) {
-                      List contentList = (List) messageMap.content;
-                      for (def cont : contentList) {
-                        if (cont.containsKey('text')) {
-                          completionText.append(cont.text);
-                        }
-                      }
-                    }
+              if (block.output?.message?.content instanceof List) {
+                for (def cont : block.output.message.content) {
+                  if (cont.containsKey('text')) {
+                    completionText.append(cont.text);
                   }
                 }
               }
-            }else if (ctx.aws_bedrock.invocation.output.output_body_json instanceof String){
+            } else if (ctx.aws_bedrock.invocation.output.output_body_json instanceof String){
               completionText.append(ctx.aws_bedrock.invocation.output.output_body_json)
             }
           }
```
Are these reasonable changes IYO?

**Review by efd6 — APPROVED** [review]
Approving with or without change in previous comment.

File: packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml (hunk 5 above)
Comment (agithomas) [reply packages/aws_bedrock/data_stream/invocation/elasticsearch/ingest_pipeline/default.yml:449]: We had discussed  and agreed that it is safer to keep the structure in the PR is safe as the shape of the output document is not consistent across various LLM text models.

**Pull request merged**